server/src/main/java/org/elasticsearch/search/profile/AbstractProfileBreakdown.java

```diff
@@ -25,34 +28,46 @@ public abstract class AbstractProfileBreakdown<T extends Enum<T>> {
     /**
      * The accumulated timings for this query node
      */
-    private final Timer[] timings;
+    private final Map<Integer, List<Timer>> timings;
     private final T[] timingTypes;
 
     /** Sole constructor. */
     public AbstractProfileBreakdown(Class<T> clazz) {
         this.timingTypes = clazz.getEnumConstants();
-        timings = new Timer[timingTypes.length];
-        for (int i = 0; i < timings.length; ++i) {
-            timings[i] = new Timer();
+        timings = new HashMap<>(this.timingTypes.length);
+        for (int i = 0; i < this.timingTypes.length; ++i) {
+            ArrayList<Timer> listOfTimers = new ArrayList<>();
+            listOfTimers.add(new Timer());
+            timings.put(i, listOfTimers);
         }
     }
 
-    public Timer getTimer(T timing) {
-        return timings[timing.ordinal()];
-    }
-
-    public void setTimer(T timing, Timer timer) {
-        timings[timing.ordinal()] = timer;
+    /**
+     * @param timingType the timing type to create a new {@link Timer} for
+     * @return a new {@link Timer} instance
+     */
+    public synchronized Timer getNewTimer(T timingType) {
+        Timer timer = new Timer();
+        timings.get(timingType.ordinal()).add(timer);
+        return timer;
     }
 
     /**
      * Build a timing count breakdown.
+     * If multiple timers where requested from different locations or threads from this profile breakdown,
+     * the approximation will contain the sum of each timers approximate time and count.
      */
-    public final Map<String, Long> toBreakdownMap() {
-        Map<String, Long> map = Maps.newMapWithExpectedSize(timings.length * 2);
-        for (T timingType : timingTypes) {
-            map.put(timingType.toString(), timings[timingType.ordinal()].getApproximateTiming());
-            map.put(timingType.toString() + "_count", timings[timingType.ordinal()].getCount());
+    public final synchronized Map<String, Long> toBreakdownMap() {
+        Map<String, Long> map = Maps.newMapWithExpectedSize(timings.keySet().size() * 2);
+        for (T timingType : this.timingTypes) {
```

Comment:
this should be something like this:
```suggestion
        this.timingTypes.forEach((timingType, timers) -> {...});
```
Its way faster to look up by index in a static array than to continually get from a map. 

I would also say, since we are using a hashmap already, why do we separate out the `ordinal` vs using the enum value as the hashmap key?